### PR TITLE
feat: Introducing Tracing with OpenTelemetry API #1537

### PR DESCRIPTION
**Pull request by jimit-j-shah**
* This change introduces Client-Side Tracing using OpenTelemetry API in Datastore Java SDK. 
* Client-Side Tracing using OpenTelemetry in the Datastore v1 Java SDK introduces the ability to trace API calls between Datastore client Datastore Backend by providing a breakdown of time spent in each phase of processing along with vital metadata.

Fixes https://github.com/googleapis/java-datastore/issues/1428 ☕️

### CODE REVIEW

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 1]
```diff
@@ -125,20 +154,57 @@ void setPrevTransactionId(ByteString transactionId) {
...
+                          io.opentelemetry.api.trace.Span.wrap(
+                              parentSpanContext.getSpanContext())));
+      return otelTraceUtil.addSettingsAttributesToCurrentSpan(spanBuilder).startSpan();
+    }
+
     @Override
     public T call() throws DatastoreException {
-      transaction = datastore.newTransaction(options);
-      try {
+      // TODO Instead of using OTel Spans directly, TraceUtil.Span should be used here. However,
```
Comment (cindy-peng) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:177]: Nit: is it possible to link the bug ID here?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 2]
```diff
@@ -99,12 +122,18 @@ static class ReadWriteTransactionCallable<T> implements Callable<T> {
     private volatile TransactionOptions options;
     private volatile Transaction transaction;
 
+    private final com.google.cloud.datastore.telemetry.TraceUtil.SpanContext parentSpanContext;
```
Comment (cindy-peng) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:125]: Can we import these as headers they appear several times in this file?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java [hunk 3]
```diff
@@ -125,20 +154,57 @@ void setPrevTransactionId(ByteString transactionId) {
...
 
+    private io.opentelemetry.api.trace.Span startSpanWithParentContext(
+        String spanName,
+        com.google.cloud.datastore.telemetry.TraceUtil.SpanContext parentSpanContext) {
+      com.google.cloud.datastore.telemetry.TraceUtil otelTraceUtil =
+          datastore.getOptions().getTraceUtil();
+      SpanBuilder spanBuilder =
+          otelTraceUtil
+              .getTracer()
+              .spanBuilder(com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN)
```
Comment (cindy-peng) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:165]: Same as above, considering import as headers

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java [hunk 4]
```diff
@@ -40,6 +39,7 @@
 import com.google.datastore.v1.RunAggregationQueryResponse;
 import com.google.datastore.v1.RunQueryRequest;
 import com.google.datastore.v1.RunQueryResponse;
+import io.opencensus.trace.Tracing;
```
Comment (cindy-peng) [inline google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java:42]: Do we need to add this?

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 1 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: This has been fixed in the latest PR so this comment is oboslete.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 2 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Not using this anymore.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java (hunk 3 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java]: Done.

File: google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java (hunk 4 above)
Comment (jimit-j-shah) [reply google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java:42]: Due to removal of the datastore.TraceUtil class, this needs to be added to continue supporting opencensus tracing for HTTP.

**Review by cindy-peng — APPROVED** [review]

**Review by ehsannas — APPROVED** [review]
LGTM. Thanks @jimit-j-shah . I haven't done an in-depth review since I'm not a Datastore expert, but I've read through the public APIs and the tracing design and they look great.

**Pull request merged**